mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -0,0 +1,819 @@
+
+impl<Coin, Storage> TransitionFrom<WaitForHistoryUpdateTrigger<Coin, Storage>>
+    for FetchingTransactionsData<Coin, Storage>
+{
+}
+
+impl<Coin, Storage> TransitionFrom<FetchingTransactionsData<Coin, Storage>>
+    for WaitForHistoryUpdateTrigger<Coin, Storage>
+{
+}
+
+#[async_trait]
+impl<Coin, Storage> State for WaitForHistoryUpdateTrigger<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        loop {
+            Timer::sleep(30.).await;
+
+            let ctx_balances = ctx.balances.clone();
+
+            let balances = match ctx.coin.all_balances().await {
+                Ok(balances) => balances,
+                Err(_) => {
+                    return Self::change_state(OnIoErrorCooldown::new(self.address.clone(), self.last_height_state));
+                },
+            };
+
+            if balances != ctx_balances {
+                // Update balances
+                ctx.balances = balances;
+
+                return Self::change_state(FetchingTransactionsData::new(
+                    self.address.clone(),
+                    self.last_height_state,
+                ));
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl<Coin, Storage> State for FetchingTransactionsData<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        const TX_PAGE_SIZE: u8 = 50;
+
+        const DEFAULT_TRANSFER_EVENT_COUNT: usize = 1;
+        const CREATE_HTLC_EVENT: &str = "create_htlc";
+        const CLAIM_HTLC_EVENT: &str = "claim_htlc";
+        const TRANSFER_EVENT: &str = "transfer";
+        const ACCEPTED_EVENTS: &[&str] = &[CREATE_HTLC_EVENT, CLAIM_HTLC_EVENT, TRANSFER_EVENT];
+        const RECIPIENT_TAG_KEY: &str = "recipient";
+        const SENDER_TAG_KEY: &str = "sender";
+        const RECEIVER_TAG_KEY: &str = "receiver";
+        const AMOUNT_TAG_KEY: &str = "amount";
+
+        struct TxAmounts {
+            total: BigDecimal,
+            spent_by_me: BigDecimal,
+            received_by_me: BigDecimal,
+        }
+
+        fn get_tx_amounts(amount: u64, sent_by_me: bool, is_self_transfer: bool) -> TxAmounts {
+            let amount = BigDecimal::from(amount);
+
+            let spent_by_me = if sent_by_me && !is_self_transfer {
+                amount.clone()
+            } else {
+                BigDecimal::default()
+            };
+
+            let received_by_me = if !sent_by_me || is_self_transfer {
+                amount.clone()
+            } else {
+                BigDecimal::default()
+            };
```

Comment:
Good catch. Thanks!

Reply:
`received_by_me` can't be 0 when tx sent by another address because we have `!sent_by_me` check as well. `received_by_me` is 0 when you send the tx but not to yourself.